### PR TITLE
[CORDA-1243]: Avoid drain-related deadlocks between 2 nodes.

### PR DESCRIPTION
**Pull request by sollecitom**
https://r3-cev.atlassian.net/browse/CORDA-1243

### CODE REVIEW

File: core/src/test/kotlin/net/corda/core/flows/FlowTestsUtils.kt [hunk 1]
```diff
@@ -49,6 +49,21 @@ inline fun <I : FlowLogic<*>, reified R : Any> StartedNode<InternalMockNetwork.M
     internalRegisterFlowFactory(initiatingFlowType.java, InitiatedFlowFactory.Core { session -> Answer(session, value) }, Answer::class.javaObjectType, true)
 }
 
+/**
+ * Returns an [InitiatingFlow] that executes the logic specified.
+ */
+fun <R> initiatingFlow(action: FlowLogic<R>.() -> R): FlowLogic<R> {
```
Comment (sollecitom) [inline core/src/test/kotlin/net/corda/core/flows/FlowTestsUtils.kt]: Unused at the moment, but still useful (developed with MockNetwork in mind).

File: node/src/integration-test/kotlin/net/corda/MessageState.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import net.corda.core.serialization.CordaSerializable
+import net.corda.core.transactions.LedgerTransaction
+import javax.persistence.Column
+import javax.persistence.Entity
+import javax.persistence.Table
+
+@CordaSerializable
+data class Message(val value: String)
+
+data class MessageState(val message: Message, val by: Party, override val linearId: UniqueIdentifier = UniqueIdentifier()) : LinearState, QueryableState {
```
Comment (sollecitom) [inline node/src/integration-test/kotlin/net/corda/MessageState.kt]: Moved out of an existing test.

File: core/src/test/kotlin/net/corda/core/flows/FlowTestsUtils.kt (hunk 1 above)
Comment (exFalso) [reply core/src/test/kotlin/net/corda/core/flows/FlowTestsUtils.kt]: This won't work with quasar. Quasar needs to instrument `call` at the point it calls `action`, but there's no annotation to indicate to it that this must be done

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt [hunk 3]
```diff
@@ -640,8 +641,9 @@ class StateMachineManagerImpl(
             }
         }
 
+        val additionalHeaders = fiber?.context?.origin.let { if (it is InvocationOrigin.Peer && it.party == party.name) { emptyMap() } else message.additionalHeaders() }
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManagerImpl.kt]: Definitely needs a comment explaining what this is and its limitation

**Pull request merged**